GoMeet/src/main/java/com/google/sps/servlets/GoogleCalendarTimesServlet.java

```diff
@@ -0,0 +1,185 @@
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/gcal-find-times")
+public class GoogleCalendarTimesServlet extends HttpServlet {
+  // The timezone that the servlet will treat PERIOD_START and PERIOD_END in
+  private static String TIMEZONE = "AET"; // Australian Eastern Time by default
+
+  // TODO: Load the API_KEY from a file.
+  private static String API_KEY = "";
+  
+  /** Field names used for accessing pairs in the query string 
+    * sent to this servlet */
+  public static class QueryStringFieldNames {
+    public static final String GUEST_LIST = "guest-list";
+    public static final String DURATION_MINS = "duration-mins";
+    public static final String DURATION_HOURS = "duration-hours";
+    public static final String PERIOD_START = "period-start";
+    public static final String PERIOD_END = "period-end";
+  }
+
+  /** 
+   * GET request for automatically proposed times using Google Calendar.
+   * Returns the list of possible meeting times in JSON format.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String guestListUri = request.getParameter(QueryStringFieldNames.GUEST_LIST);
+    String durationMinsStr = request.getParameter(QueryStringFieldNames.DURATION_MINS); 
+    String durationHoursStr = request.getParameter(QueryStringFieldNames.DURATION_HOURS);
+    String periodStartStr = request.getParameter(QueryStringFieldNames.PERIOD_START); 
+    String periodEndStr = request.getParameter(QueryStringFieldNames.PERIOD_END);
+
+    if (guestListUri == null || durationHoursStr == null || durationMinsStr == null || 
+        periodStartStr == null || periodEndStr == null) {
+      ServletUtil.sendErrorResponse(response, 
+          HttpServletResponse.SC_BAD_REQUEST, 
+          ErrorMessages.BAD_REQUEST_ERROR);
+      return;
+    }
+
+    // 'guestListStr' are sent by the request as an encoded URI.
+    String guestListStr = ServletUtil.decodeUri(guestListUri);
+    // Each element in the decoded list is deliminated by a comma.
+    ArrayList<String> guestList = new ArrayList<String> (Arrays.asList(guestListStr.split(","))); 
+
+    // Convert hours and minutes duration to milliseconds.
+    int totalDurationMs = 
+        (Integer.parseInt(durationHoursStr) * 60 + Integer.parseInt(durationMinsStr)) * 60 * 1000;
+    
+    Date periodStart;
+    Date periodEnd;
+    
+    try {
+      SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm");
+      formatter.setTimeZone(TimeZone.getTimeZone(TIMEZONE));
+      periodStart = formatter.parse(periodStartStr);
+      periodEnd = formatter.parse(periodEndStr);
+    } catch (ParseException e) {
+      ServletUtil.sendErrorResponse(response, 
+          HttpServletResponse.SC_BAD_REQUEST, 
+          ErrorMessages.INVALID_ARG_FORMAT);
```

Comment:
I presume that periodStartStr and periodEndStr are both generated by the front end and do not come from the user? (and, I suppose the frontend is validating anything the user enters anyway?)

Just my opinion, but If so, from a user perspective, I wonder what benefit it adds to tell them that the argument format is invalid rather than say, the backend got a bad request like on line 63? The backend could log the problem so that a sysadmin/developer could come in and see what's going on, but I suspect that from the user's point of view, they don't really have control over the argument format.

Reply:
Yep, both the strings are generated by front end code. I was thinking the more specific error message might help identify which part of the query string was 'bad', which might help with debugging the front end code. It wasn't intended as a message for users, more for the front end developer. 

But I guess it would be justifiable to just send a generic error message - developer will still get a reasonable understanding of what went wrong. Or, I could also just throw the ParseException error rather than sending an error response - either way, the developer will get the message that there is a backend error.

Reply:
btw changing this is optional.

Reply:
Yep, OK 👍